src/java.base/share/classes/java/lang/reflect/code/analysis/StringConcatTransformer.java

```diff
@@ -38,48 +40,78 @@
+        // Root of concat tree, zero uses, two or more uses,
+        // or one use that is not a subsequent concat op
+        Set<Op.Result> uses = cz.result().uses();
+        return uses.size() != 1 || !(uses.iterator().next().op() instanceof CoreOp.ConcatOp);
+    }
 
-            Value result = cz.result();
+    static void buildFromTree(Block.Builder block, Op.Result builder, CoreOp.ConcatOp cz) {
+        // Process concat op's operands from left to right
+        buildFromTree(block, builder, cz.operands().get(0));
+        buildFromTree(block, builder, cz.operands().get(1));
+    }
 
-            StringAndBuilder newRes = stringBuild(builder, left, right);
-            builder.context().mapValue(result, newRes.string);
+    static void buildFromTree(Block.Builder block, Op.Result builder, Value v) {
+        if (v instanceof Op.Result r &&
+                r.op() instanceof CoreOp.ConcatOp cz &&
+                r.uses().size() == 1) {
+            // Node of tree, recursively traverse the operands
+            buildFromTree(block, builder, cz);
         } else {
-            builder.op(op);
+            // Leaf of tree, append value to builder
+            // Note leaf can be the result of a ConcatOp with multiple uses
+            block.op(append(block, builder, block.context().getValue(v)));
         }
-        return builder;
     }
 
-    //Uses StringBuilder and Immediate String Value
-    private static StringAndBuilder stringBuild(Block.Builder builder, Value left, Value right) {
-        var newB = stringBuilder(builder, left, right);
-        var toStringInvoke = CoreOp.invoke(SB_TO_STRING, newB);
-        Value newString = builder.apply(toStringInvoke);
-        return new StringAndBuilder(newString, newB);
+    private static Op append(Block.Builder block, Value builder, Value arg) {
+        return append(block, builder, arg, arg.type());
     }
 
-    private static Value stringBuilder(Block.Builder builder, Value left, Value right) {
-        CoreOp.NewOp newBuilder = CoreOp._new(FunctionType.functionType(SBC_TYPE));
-        Value sb = builder.apply(newBuilder);
-        builder.op(append(sb, left));
-        builder.op(append(sb, right));
-        return sb;
+    private static Op append(Block.Builder block, Value builder, Value arg, TypeElement type) {
+        // Check if we need to widen unsupported integer types in the StringBuilder API
+        // Strings are fed in as-is, everything else given as an Object.
+        TypeElement appendType = type;
+        Value appendValue = arg;
+        if (type instanceof PrimitiveType || type.equals(JavaType.J_L_STRING)) {
+            //Widen if we need to.
+            if (type.equals(JavaType.BYTE) || type.equals(JavaType.SHORT)) {
+                appendValue = block.op(CoreOp.conv(JavaType.INT, arg));
+                appendType = JavaType.INT;
+            }
+        }
+        else {
+            appendType = JavaType.J_L_OBJECT;
+        }
```

Comment:
I think you can simplify as its hard to follow the logic when the type is string. We can now collapse the append methods into one (the two methods were used in a prior version).
e.g.,
```
private static Op append(Block.Builder block, Value builder, Value arg) {
   TypeElement type = arg.type();
   if (type.equals(JavaType.BYTE) || type.equals(JavaType.SHORT)) {
       // Widen
       arg = block.op(CoreOp.conv(JavaType.INT, arg));
       type = JavaType.INT;
    } else if (!type.equals(JavaType.J_L_STRING)) {
        type = JavaType.J_L_OBJECT;
    }

    MethodRef appendMethod = MethodRef.method(J_L_STRING_BUILDER, "append", J_L_STRING_BUILDER, type);
    return CoreOp.invoke(appendMethod, builder, arg)
}
```

Reply:
Good move. Done.